src/shared/Microsoft.Git.CredentialManager/GitConfiguration.cs

```diff
@@ -34,10 +45,11 @@ public interface IGitConfiguration
         /// Try and get the value of a configuration entry as a string.
         /// </summary>
         /// <param name="level">Filter to the specific configuration level.</param>
+        /// <param name="type">Type constraint to which the config value should be canonicalized.</param>
         /// <param name="name">Configuration entry name.</param>
         /// <param name="value">Configuration entry value.</param>
         /// <returns>True if the value was found, false otherwise.</returns>
-        bool TryGet(GitConfigurationLevel level, string name, out string value);
+        bool TryGet(GitConfigurationLevel level, GitConfigurationType type, string name, out string value);
```

Comment:
I think it might be better/less noisy to put this `GitConfigurationType type` at the end of every existing method's argument list (except `out` params), i.e..

```suggestion
        bool TryGet(GitConfigurationLevel level, string name, GitConfigurationType type, out string value);
```

..and then add appropriate extension methods to the `GitConfigurationExtensions` class that pass the default `GitConfigurationType.None` in as this extra argument.

This means that existing call-sites do not need to be updated (they'll end up resolving to these new extension methods with the old arg lists instead), and for call-sites that _do_ want to also specify a coerced type, it's just adding one more argument to the end of their call.

Reply:
I had considered that (specifically, adding the argument at the end and having it use `GitConfigurationExtensions.None` by default), but I don't necessarily find it cleaner. There are only 5 non-test usages outside of `GitConfiguration` itself ([1](https://github.com/microsoft/Git-Credential-Manager-Core/pull/439/commits/7d644a4a29b3c15b90682aebc551067ba0def3ed#diff-62a94bca9b60857fb75038ef1cd82ecfc5dc87699a88a31cfadfb62e075261eaL61-R62), [2 & 3](https://github.com/microsoft/Git-Credential-Manager-Core/pull/439/commits/7d644a4a29b3c15b90682aebc551067ba0def3ed#diff-0be9b83a69df29220151264feb5c7fd3822f1a0ba4d9463885bd92bd772f6a60L97-R100), [4](https://github.com/microsoft/Git-Credential-Manager-Core/pull/439/commits/7d644a4a29b3c15b90682aebc551067ba0def3ed#diff-76e78a4449281f62e3fecf0bb1e92425e9bf4e6d031c6e26ffaa3ce9c12974a4L170-R170), [5](https://github.com/microsoft/Git-Credential-Manager-Core/pull/439/commits/7d644a4a29b3c15b90682aebc551067ba0def3ed#diff-76e78a4449281f62e3fecf0bb1e92425e9bf4e6d031c6e26ffaa3ce9c12974a4L221-R221)) of the non-extension `TryGet` & `GetAll` that needed updating, and updating them here is a one-time change. Conversely, an extension method for just those 5 cases is a persistent addition, one that as far as I can tell is just a way of hiding one of two optional parameters (as opposed to the current configuration, which I interpret as "the full method with an extension for hiding defaulting behavior").

One thing this made me realize, though, is that adding more and more options over time could get unwieldy (either with lots of extensions or long function signatures). A `GitConfigOptions` structure (initialized with default values for `GitConfigurationLevel` and `GitConfigurationType`) might be a clean way of handling that and future-proofing it for other options. What do you think?

Reply:
> There are only 5 non-test usages outside of GitConfiguration itself (1, 2 & 3, 4, 5) of the non-extension TryGet & GetAll that needed updating, and updating them here is a one-time change.

A fair enough argument! 😁 

One final suggestion then.. I read `GitConfigurationType.None` strangely.. I mean, we're not getting a "none" type, but rather we're not coalescing the raw data in to a specific type..

Perhaps `GitConfigurationType.String` or `GitConfigurationType.Raw` would be more suitable, IMO.

Reply:
Makes sense! Updated to `GitConfigurationType.Raw`.